opt: include NULL first column values in spans between partitions

mgartner opened the pull request:
Previously, index spans which were meant to include values in-between
index partition values did not include NULL values for the first index
column. As a result, a partitioned index scan did not produce any rows
where the value of the first index column was NULL. This commit fix the
bug by including NULL for the first index column in the index spans if
the column is nullable.

Fixes #63733

Release note (bug fix): A correctness bug, which caused partitioned
index scans to omit rows where the value of the first index column was
NULL, has been fixed. This bug was present since v19.2.0.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/63932)
<!-- Reviewable:end -->

RaduBerinde reviewed (commented):
Argh, those damn NULLs!

I'm struggling to understand why this is only a problem for the first column.. If we synthesize a filter like `((col1, col2) < ('foo', 'bar') OR (col1, col2) > ('fud', 'baz'))`, isn't `('foo', NULL)` also problematic?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/63932#-:-MYkCbpR4Pd5AoaetBxR:bg4490g)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @rytaft)



<!-- Sent from Reviewable.io -->

mgartner commented:
It's not a problem because the span for the `(col1, col2) < ('foo', 'bar')` part of the expression would be `[/NULL - /'foo'/'bar')` and the key `/'foo'/NULL` lies within that span. When we build index constraints for tuple inequalities we specifically exclude NULL for the first column because the inequality will always evaluate to NULL in that case: https://github.com/cockroachdb/cockroach/blob/09be4f72a054cf66ef603b1e4525844c9ef547fc/pkg/sql/opt/idxconstraint/index_constraints.go#L454-L455

But if subsequent values in the tuple are NULL, the inequality could be `true` or `false` in some cases, so we can't exclude NULLs for those columns during the scan.

```
SELECT (1, null) < (1, 2);
  ?column?
------------
    NULL

SELECT (1, null) < (2, 2);
  ?column?
------------
    true

SELECT (1, null) > (2, 2);
  ?column?
------------
   false
```





<!-- Sent from Reviewable.io -->

mgartner reviewed (commented):
I updated the code comment to briefly explain why this is only required for the first column in the index.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/63932#-:-MYkOOMc0tp3hLxtQQpa:b-7lswz6)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @rytaft)



<!-- Sent from Reviewable.io -->

RaduBerinde commented:
The resulting spans are ok, but it is all super fragile. It breaks the conceptual model where `inBetweenFilters` is supposed to return filters that don't affect the results. We rely on specifics on how `tryConstrainIndex` handles the optional filters (and that their remaining filters don't make it into `inBetweenRemainingFilters`). At the very least we should document these subtleties in the relevant methods.

Also, is it possible to have a NULL value on the right hand side?

mgartner commented:
> It breaks the conceptual model where inBetweenFilters is supposed to return filters that don't affect the results.

I don't understand. `inBetweenFilters` is required for generating correctly constrained partitioned index scans.

RaduBerinde commented:
> I don't understand. inBetweenFilters is required for generating correctly constrained partitioned index scans.

I mean they are filters that conceptually you could apply on top of the scan and they wouldn't filter out any results..  How else do you define an "optional filter"?

mgartner commented:
I understand that, but the essence of `inBetweenFilters` doesn't break this model, the bug in it did. Or is there another reason that `inBetweenFilters` fundamentally breaks this contract that I'm not seeing.

mgartner commented:
Well I guess `inBetweenFilters` on their own would filter out rows - it has to be combined with the partition value filters. I agree it's all fragile.

RaduBerinde commented:
We can look at it from another direction - how should we document `optionalFilters` for `tryConstrainIndex`? To me those are filters that we know are true for all resulting rows (after you applied `requiredFilters`), so you can take them or leave them as needed to best constrain the index.

This is not true of `(col1, col2) < ('foo', 'bar')`, because it incorrectly excludes `('foo', NULL)`. It is in principle conceivable that the index constrain code could use this optional filter to infer that `('foo', NULL)` is excluded and remove it from whatever spans it generates - the tuple code just happens to not do that currently.

By the way, in general this would be an improvement to the tuple index constraints code: for `(a,b) < (1,2)`, it would be better to have spans `[/NULL - /1) (/1/NULL - /1/2)` than `(/NULL - /1/2]` with an extra remaining filter. If we were to make this improvement, it would break the fix in this PR, no?

I think this fix is adequate for a backport, but we should 1) document these subtleties best we can and 2) think about how to improve it on master. Maybe a clean solution would be to introduce separate tuple comparison operators which are like regular comparison operators but with "simple" null handling (where NULL equals NULL and NULL is smaller than any non-NULL). Or, perhaps we can implement these in-between spans in a different way that doesn't require actual filter expressions to be synthesized.

mgartner committed:
opt: include NULL first column values in spans between partitions

Previously, index spans which were meant to include values in-between
index partition values did not include NULL values for the first index
column. As a result, a partitioned index scan did not produce any rows
where the value of the first index column was NULL. This commit fix the
bug by including NULL for the first index column in the index spans if
the column is nullable.

Fixes #63733

Release note (bug fix): A correctness bug, which caused partitioned
index scans to omit rows where the value of the first index column was
NULL, has been fixed. This bug was present since v19.2.0.

RaduBerinde reviewed (commented):
Since Marcus is out and a customer wants a fix, I will merge and backport the PR. I have added a TODO along the lines of my complaint above. @rytaft, could you do a final review? Thanks!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/63932#-:-MYqHOR9AloROsx-5FeU:bauaj5d)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @rytaft)



<!-- Sent from Reviewable.io -->

rytaft reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 3 of 3 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/63932#-:-MYqSpA56QJM1USqNnGO:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @mgartner)

___
*[pkg/sql/opt/xform/select_funcs.go, line 492 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/63932#-MYqS2QbAUF3y4vCQ_BY:-MYqS2QbAUF3y4vCQ_BZ:b-sgjpph) ([raw file](https://github.com/cockroachdb/cockroach/blob/d247dc9a53f07423c03292afbf9db3c1e9d3bea2/pkg/sql/opt/xform/select_funcs.go#L492)):*
> ```Go
> 	// This is only required for the first column in the index because the index
> 	// constraint spans built for tuple inequalities only exclude NULL values
> 	// for the first element in the tuple.
> ```

seems like it would be good to add a test where the second (or third) column is also nullable and show that we return null values in that case


<!-- Sent from Reviewable.io -->

RaduBerinde reviewed (commented):
TFTR!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/63932#-:-MYrEGY0DhNkkzJtv8v6:b-814g8k)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @rytaft)

___
*[pkg/sql/opt/xform/select_funcs.go, line 492 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/63932#-MYqS2QbAUF3y4vCQ_BY:-MYrEIhaBk1p1IyzJU6M:b-48vilg) ([raw file](https://github.com/cockroachdb/cockroach/blob/d247dc9a53f07423c03292afbf9db3c1e9d3bea2/pkg/sql/opt/xform/select_funcs.go#L492)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

seems like it would be good to add a test where the second (or third) column is also nullable and show that we return null values in that case
</blockquote></details>

Good call, done.


<!-- Sent from Reviewable.io -->

rytaft reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 2 of 2 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/63932#-:-MYrTobf3mkgUprY-KfX:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @mgartner)



<!-- Sent from Reviewable.io -->

RaduBerinde commented:
bors r+
